protocols/kad/src/behaviour.rs

```diff
@@ -2701,21 +2751,26 @@ impl GetClosestPeersError {
 /// The result of [`Kademlia::get_providers`].
 pub type GetProvidersResult = Result<GetProvidersOk, GetProvidersError>;
 
-/// The successful result of [`Kademlia::get_providers`].
+/// The successful progress result of [`Kademlia::get_providers`].
 #[derive(Debug, Clone)]
-pub struct GetProvidersOk {
+pub struct GetProvidersProgress {
     pub key: record::Key,
-    pub providers: HashSet<PeerId>,
+    pub provider: PeerId,
```

Comment:
Why return them one by one? Why not as a batch? One always discovers them in batches, no?

Reply:
Yeah, I went back and for on this one. The nice thing about this, is that I can reuse the sequence number as the count of results, and the processing is a little more streamlined. But batching is definitely more efficient given they come in as batches.

Reply:
Do we have any idea about the difference in the performance profile?

From an API PoV, it seems nicer to return them one-by-one given that we are already returning a list of events now. For example, if I want to write a log line for each progress entry, being given a set makes this slightly more cumbersome.

Reply:
> Do we have any idea about the difference in the performance profile?

I don't know, the cost is `1` vs `n` events per discovery. From my understanding that cost should be quite low, but I might be mistaken.

Reply:
Intuitively I am also assuming that neither has a performance benefit.

Returning them one by one does loose information. Returning them batched gives the user the information that all records within the same batch came from the same source.

Reply:
@mxinden updated, please check out